egraph support: rewrite to work in terms of CLIF data structures.

cfallin opened the pull request:
This work rewrites the "egraph"-based optimization framework in
Cranelift to operate on aegraphs (acyclic egraphs) represented in the
CLIF itself rather than as a separate data structure to which and from
which we translate the CLIF.

The basic idea is to add a new kind of value, a "union", that is like an
alias but refers to two other values rather than one.  This allows us to
represent an eclass of enodes (values) as a tree. The union node allows
for a value to have *multiple representations*: either constituent value
could be used, and (in well-formed CLIF produced by correct
optimization rules) they must be equivalent.

Like the old egraph infrastructure, we take advantage of acyclicity and
eager rule application to do optimization in a single pass. Like before,
we integrate GVN (during the optimization pass) and LICM (during
elaboration).

Unlike the old egraph infrastructure, everything stays in the
DataFlowGraph. "Pure" enodes are represented as instructions that have
values attached, but that are not placed into the function layout. When
entering "egraph" form, we remove them from the layout while optimizing.
When leaving "egraph" form, during elaboration, we can place an
instruction back into the layout the first time we elaborate the enode;
if we elaborate it more than once, we clone the instruction.

The implementation performs two passes overall:

- One, a forward pass in RPO (to see defs before uses), that (i) removes
  "pure" instructions from the layout and (ii) optimizes as it goes. As
  before, we eagerly optimize, so we form the entire union of optimized
  forms of a value before we see any uses of that value. This lets us
  rewrite uses to use the most "up-to-date" form of the value and
  canonicalize and optimize that form.

  The eager rewriting and acyclic representation make each other work
  (we could not eagerly rewrite if there were cycles; and acyclicity
  does not miss optimization opportunities only because the first time
  we introduce a value, we immediately produce its "best" form). This
  design choice is also what allows us to avoid the "parent pointers"
  and fixpoint loop of traditional egraphs.

  This forward optimization pass keeps a scoped hashmap to "intern"
  nodes (thus performing GVN), and also interleaves on a per-instruction
  level with alias analysis. The interleaving with alias analysis allows
  alias analysis to see the most optimized form of each address (so it
  can see equivalences), and allows the next value to see any
  equivalences (reuses of loads or stored values) that alias analysis
  uncovers.

- Two, a forward pass in domtree preorder, that "elaborates" pure enodes
  back into the layout, possibly in multiple places if needed. This
  tracks the loop nest and hoists nodes as needed, performing LICM as it
  goes. Note that by doing this in forward order, we avoid the
  "fixpoint" that traditional LICM needs: we hoist a def before its
  uses, so when we place a node, we place it in the right place the
  first time rather than moving later.

This PR replaces the old (a)egraph implementation. It removes both the
cranelift-egraph crate and the logic in cranelift-codegen that uses it.

On `spidermonkey.wasm` running a simple recursive Fibonacci
microbenchmark, this work shows 5.5% compile-time reduction and 7.7%
runtime improvement (speedup). More benchmarks to come.

Most of this implementation was done in (very productive) pair
programming sessions with Jamey Sharp, thus:

Co-authored-by: Jamey Sharp <jsharp@fastly.com>

bjorn3 commented on cranelift/codegen/Cargo.toml:
```diff
@@ -29,6 +28,7 @@ smallvec = { workspace = true }
 regalloc2 = { version = "0.5.0", features = ["checker"] }
 souper-ir = { version = "2.1.0", optional = true }
 sha2 = { version = "0.10.2", optional = true }
+fxhash = "0.2.1"
```
We already have an fxhash impl in the fx module.

bjorn3 commented:
How are alias values printed in the textual format of clif ir? And can it be read back?

bjorn3 commented on cranelift/codegen/Cargo.toml:
```diff
@@ -18,8 +18,7 @@ bumpalo = "3"
 cranelift-bforest = { workspace = true }
-cranelift-egraph = { workspace = true }
-hashbrown = { workspace = true, optional = true }
+hashbrown = { workspace = true }
```
What exactly is hashbrown necessary for? std::collection::HashMap is a wrapper around hashbrown.

cfallin replied:
It's necessary for `crate::ctxhash::CtxHashMap`, which is a hashmap that uses external context for hashing and equality. `HashMap` isn't sufficient for this.

Note that this was previously in the `cranelift-egraph` crate, which was unconditionally required by `cranelift-codegen`, and now has been moved into `cranelift-codegen`; so there's no change in the dependency footprint.

cfallin commented:
> How are alias values printed in the textual format of clif ir? And can it be read back?

They currently aren't; because union values aren't present in the layout, the usual CLIF writer won't see them (it ignores all insts and values not "in the function" as per the layout). I agree it would be great to come up with a text format for egraph-stage CLIF; maybe as a followup to the initial PR?

fitzgen commented on cranelift/codegen/src/context.rs:
```diff
@@ -196,10 +196,16 @@ impl Context {
             self.compute_loop_analysis();
-            let mut eg = FuncEGraph::new(&self.func, &self.domtree, &self.loop_analysis, &self.cfg);
-            eg.elaborate(&mut self.func);
+            let mut alias_analysis = AliasAnalysis::new(&self.func, &self.domtree);
+            let mut pass = EgraphPass::new(
+                &mut self.func,
+                &self.domtree,
+                &self.loop_analysis,
+                &mut alias_analysis,
+            );
+            pass.run();
+            log::info!("egraph stats: {:?}", pass.stats);
```
Can we package this up in a `self.egraph_pass()` method, similar to other passes, rather than have all the bits of state exposed in our pass-ordering / mid end? I think that would be a bit nicer.

cfallin replied:
Done!

fitzgen commented on cranelift/codegen/src/egraph.rs:
```diff
@@ -1,342 +1,437 @@
+        while let Some(block) = block_stack.pop() {
+            // We popped this block; push children
+            // immediately, then process this block.
+            for child in self.domtree_children.children(block) {
+                block_stack.push(child);
+            }
```
This is shorter and allows for reserving capacity for all children at once:

```suggestion
            block_stack.extend(self.domtree_children.children(block));
```

fitzgen commented on cranelift/codegen/src/egraph.rs:
```diff
@@ -1,342 +1,437 @@
+                } else {
+                    self.stats.skeleton_inst += 1;
+
+                    // Not pure, but may still be a load or store:
+                    // process it to see if we can optimize it.
+                    if let Some(new_result) = self.alias_analysis.process_inst(
+                        cursor.func,
+                        &mut alias_analysis_state,
+                        inst,
+                    ) {
+                        self.stats.alias_analysis_removed += 1;
+                        let result = cursor.func.dfg.inst_results(inst)[0];
+                        value_to_opt_value[result] = new_result;
+                        cursor.remove_inst_and_step_back();
+                    } else {
+                        // Set all results to identity-map to themselves
+                        // in the value-to-opt-value map.
+                        for &result in cursor.func.dfg.inst_results(inst) {
+                            value_to_opt_value[result] = result;
+                            self.eclasses.add(result);
```
Could we wrap this up into `process_skeleton_inst` or something like that, just to make the overall algorithm a little easier to read at a glance?

bjorn3 commented:
> > How are alias values printed in the textual format of clif ir? And can it be read back?
> 
> They currently aren't; because union values aren't present in the layout, the usual CLIF writer won't see them (it ignores all insts and values not "in the function" as per the layout). I agree it would be great to come up with a text format for egraph-stage CLIF; maybe as a followup to the initial PR?

Sure.

fitzgen reviewed (approved):
LGTM with all my feedback addressed. Really excited for this!

cfallin commented on cranelift/codegen/src/ctxhash.rs:
```diff
@@ -0,0 +1,161 @@
+pub trait CtxHash<Value: ?Sized>: CtxEq<Value, Value> {
+    /// Compute the hash of `value`, given the context in `self` and
+    /// the union-find data structure `uf`.
+    fn ctx_hash(&self, value: &Value) -> u64;
```
No particular reason, just expediency at the time (it was this way in the original egraph prototype then carried over). Updated as suggested!

cfallin commented on cranelift/codegen/src/ctxhash.rs:
```diff
@@ -0,0 +1,161 @@
+    }
+
+    /// Look up a key, returning a borrow of the value if present.
+    pub fn get<'a, Q, Ctx: CtxEq<K, Q> + CtxHash<Q> + CtxHash<K>>(
```
Done!

cfallin commented on cranelift/codegen/src/egraph.rs:
```diff
@@ -1,342 +1,437 @@
+impl NewOrExistingInst {
+    fn get_inst_key<'a>(&'a self, dfg: &'a DataFlowGraph) -> (Type, InstructionData) {
+        match self {
+            NewOrExistingInst::New(data, ty) => (*ty, data.clone()),
```
Done!

cfallin commented on cranelift/codegen/src/egraph.rs:
```diff
@@ -1,342 +1,437 @@
+                ValueDef::Result(i, 0) => {
+                    trace!("  -> {} = {:?}", i, self.func.dfg[i]);
+                }
+                _ => {}
```
Good point, added the conditional to be sure (it never hurts to make LLVM's life easier in this way either...).

cfallin commented on cranelift/codegen/src/egraph/cost.rs:
```diff
@@ -0,0 +1,69 @@
+
+use crate::ir::Opcode;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub(crate) struct Cost(u32);
+impl Cost {
+    pub(crate) fn at_level(&self, loop_level: usize) -> Cost {
+        let loop_level = std::cmp::min(2, loop_level);
+        let multiplier = 1u32 << ((10 * loop_level) as u32);
+        Cost(self.0.saturating_mul(multiplier)).finite()
+    }
+
+    pub(crate) fn infinity() -> Cost {
+        // 2^32 - 1 is, uh, pretty close to infinite... (we use `Cost`
+        // only for heuristics and always saturate so this suffices!)
+        Cost(u32::MAX)
+    }
+
+    pub(crate) fn zero() -> Cost {
+        Cost(0)
+    }
+
+    /// Clamp this cost at a "finite" value. Can be used in
+    /// conjunction with saturating ops to avoid saturating into
+    /// `infinity()`.
+    fn finite(self) -> Cost {
+        Cost(std::cmp::min(u32::MAX - 1, self.0))
+    }
+}
```
Added! Hopefully the explanation helps.

cfallin commented on cranelift/codegen/src/loop_analysis.rs:
```diff
@@ -35,9 +35,9 @@ struct LoopData {
 /// A level in a loop nest.
 #[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
-pub struct LoopLevel(u8);
+pub struct LoopLevel(u32);
```
Ah, reverted, thanks for catching this!

For the curious, this change came from a dead-end (since reverted) where we invented a notion of a "pseudo-loop nest" based on the domtree (basic idea is that we wanted the loop nest to fully be a subtree of the domtree; consider e.g. a loop exit dominating the rest of the function, that's where the loop tree and domtree differ) and loop nests could get much deeper. But we've reverted all that and `LoopLevel` should correspond only to true loop-nest level again, now.

cfallin commented:
Updated and addressed everything, I think! (Left the one thread about multi-result instructions open; that's something we'll want to address somehow eventually, but I think isn't a quick/easy thing to address in this PR.) Thanks for the review!

cfallin commented:
Interesting benchmark discovery:

- This PR's work branched from main on Nov 19, which was between #5231 landing and #5335 landing, i.e., when we accidentally had turned on explicit bounds-checks unconditionally;
- On the `egraph-in-dfg` branch in my fork (this PR's original branch), I continue to see the speedups above;
- On the rebase (`egraph-in-dfg-rebase` in my fork), the speedups disappear into the noise.

This seems to indicate to me that on worse code (explicit bounds checks), this approach is a little more efficient, while otherwise it's more-or-less equivalent.

Given the two other benefits of flexibility (writing rewrite rules in a simple way) and verifiability, on balance I suspect this will still make sense to turn on by default once fuzz-clean. (I'm doing a followup PR to add it to fuzzing configs.) But I'm very open to discussion on that; one alternative is to keep adding rewrite rules until we see a very clear code-quality (runtime) win.

bjorn3 commented:
I just did some benchmarking of egraphs and the perf improvement is huge on the benchmark I tried:

```
Benchmark 1: ./raytracer_cg_clif
  Time (mean ± σ):      8.553 s ±  0.010 s    [User: 8.539 s, System: 0.014 s]
  Range (min … max):    8.543 s …  8.568 s    10 runs
 
Benchmark 2: ./raytracer_cg_clif_egraph
  Time (mean ± σ):      6.068 s ±  0.017 s    [User: 6.057 s, System: 0.011 s]
  Range (min … max):    6.047 s …  6.108 s    10 runs
 
Benchmark 3: ./raytracer_cg_clif_release
  Time (mean ± σ):      6.450 s ±  0.021 s    [User: 6.439 s, System: 0.012 s]
  Range (min … max):    6.410 s …  6.482 s    10 runs
 
Benchmark 4: ./raytracer_cg_clif_release_egraph
  Time (mean ± σ):      5.853 s ±  0.053 s    [User: 5.841 s, System: 0.012 s]
  Range (min … max):    5.779 s …  5.908 s    10 runs
 
Summary
  './raytracer_cg_clif_release_egraph' ran
    1.04 ± 0.01 times faster than './raytracer_cg_clif_egraph'
    1.10 ± 0.01 times faster than './raytracer_cg_clif_release'
    1.46 ± 0.01 times faster than './raytracer_cg_clif'
```

(release uses cranelift's speed_and_size opt_level as well as mir inlining and other rustc mir optimizations)
